src/wbase/Task.cc

```diff
@@ -125,39 +105,73 @@ std::atomic<uint32_t> taskSequence{0};
 /// available to define the action to take when this task is run, so
 /// Command::setFunc() is used set the action later. This is why
 /// the util::CommandThreadPool is not called here.
-Task::Task(TaskMsgPtr const& t, std::string const& query, int fragmentNumber,
-           std::shared_ptr<SendChannelShared> const& sc)
-        : msg(t),
+Task::Task(TaskMsgPtr const& t, int fragmentNumber, std::shared_ptr<UserQueryInfo> const& userQueryInfo,
+           size_t templateId, int subchunkId, std::shared_ptr<SendChannelShared> const& sc)
+        : _userQueryInfo(userQueryInfo),
           _sendChannel(sc),
           _tSeq(++taskSequence),
           _qId(t->queryid()),
+          _templateId(templateId),
+          _hasChunkId(t->has_chunkid()),
+          _chunkId(t->has_chunkid() ? t->chunkid() : -1),
```

Comment:
As a general comment, I think that negative numbers make no sense for in-chunk and sub-chunk identifiers which should be modeled as natural numbers: 0, 1, 2, etc. Chunks/sub-chunks aren't meant to be used in arithmetic operations (addition or substractions where the sign is needed). And yes, I'm aware that the current Protobuf definition for these attributes incorrectly defines those as signed integers. I think that was a mistake. 

Apart from that statement, my specific comment in this context is that you have already introduced the special flag `_hasChunkId` that allows tracking the state of the `chunkId` attribute.  Similar actions should be also applied to other optional attributes that you're fetching from the Protobuf messages and storing within tasks. If you do this for all optional attributes then your code would be more consistent and safer.

Besides, you could use those `has` states to validate attributes in the comparison operations. See my comments posted on this subject for operations `ChunkIdGreater` and `ChunkIdEqual`.

Reply:
I disagree as an invalid chunkId or subchunkId in the middle of a table or column name should never happen, but if it does '-1' will cause a SQL parse error, where '0' would could go through in some cases. I believe '0' is a valid id.